sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util.implementation;
+
+import com.azure.core.util.ReferenceManager;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Implementation of {@link ReferenceManager}.
+ */
+public final class ReferenceManagerImpl implements ReferenceManager, Runnable {
+    // Used to create a unique thread which is used to clean up the resources managed by an instance ReferenceManager.
+    private static final AtomicInteger RESOURCE_MANAGER_THREAD_NUMBER = new AtomicInteger();
+
+    // Base name for ResourceManager threads.
+    private static final String BASE_THREAD_NAME = "azure-sdk-referencemanager-";
+
+    private final CleanableReference<?> cleanableReferenceList;
+    private final ReferenceQueue<Object> queue;
+
+    /**
+     * Creates a new instance of {@link ReferenceManagerImpl}.
+     */
+    public ReferenceManagerImpl() {
+        this.queue = new ReferenceQueue<>();
+        this.cleanableReferenceList = new CleanableReference<>();
+
+        new CleanableReference<>(this, () -> { }, this);
```

Comment:
I would recommend a comment for this line, because as it is it looks very odd. You should get into the habit of talking to yourself in code like this, to explain why it is this way.